model_zoo/pytorch/nanogpt/train_lightning.py

```diff
@@ -0,0 +1,387 @@
+# Copyright 2023 The DLRover Authors. All rights reserved. Licensed
+# under the Apache License, Version 2.0 (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+import argparse
+import contextlib
+import math
+import os
+import pickle
+from datetime import timedelta
+
+import numpy as np
+import pytorch_lightning as pl
+import torch
+import torch.distributed as dist
+from model import GPT, GPTConfig
+from torch.utils.data import DataLoader, Dataset
+
+from dlrover.trainer.torch.elastic_sampler import ElasticDistributedSampler
+
+local_rank = None
+master_process = False
+meta_vocab_size = None
+
+
+class UpdateDataStepCallback(pl.Callback):
+    def on_train_batch_end(
+        self,
+        trainer,
+        pl_module,
+        outputs,
+        batch,
+        batch_idx,
+    ) -> None:
+        trainer.datamodule.global_step = trainer.global_step
+
+    def on_train_epoch_start(self, trainer, pl_module) -> None:
+        trainer.datamodule.train_sampler.set_epoch(trainer.current_epoch)
+
+
+class Nanogpt(pl.LightningModule):
+    global meta_vocab_size
```

Comment:
Why do we use a global variable?

Reply:
Since we need to ensure the consistency of the meta_vocab_size of the model and the meta_vocab_size of the training data. For pretraining , the meta_vocab_size is determined by training dataset. For convenience, I used a global variable to pass meta_vocab_size.

Reply:
Alternatively, we should load meta_vocab_size from the model meta config or from the dataset config by adding a `load_meta_vocab_size()` function.

Reply:
Modified the `_get_meta_vocab_size()` to remove global varible.